### PR TITLE
[SPARK-19634][ML] Multivariate summarizer - dataframes API

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

This patch adds the DataFrames API to the multivariate summarizer (mean, variance, etc.). In addition to all the features of MultivariateOnlineSummarizer, it also allows the user to select a subset of the metrics. 

## How was this patch tested?

Testcases added.

## Performance
Resolve several performance issues in #17419, further optimization pending on SQL team's work. One of the SQL layer performance issue related to these feature has been resolved in #18712, thanks @liancheng and @cloud-fan 

### Performance data

(test on my laptop, use 2 partitions. tries out = 20, warm up = 10)

The unit of test results is records/milliseconds (higher is better)

Vector size/records number | 1/10000000 | 10/1000000 | 100/1000000 | 1000/100000 | 10000/10000
----|------|----|---|----|----
Dataframe | 15149  | 7441 | 2118 | 224 | 21
RDD from Dataframe | 4992  | 4440 | 2328 | 320 | 33
raw RDD | 53931  | 20683 | 3966 | 528 | 53

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: The `SummarizerBuffer` is similar to `MultivariateOnlineSummarizer`. But it has some new features:
1) support computing only part of the metrics, saving the buffer memory cost. 
2) support a optimized input interface, take advantage of saving data copy.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Here I do not use `VectorUDT.deserialize` but directly manipulate the `UnsafeArrayData` coming from `InternalRow`(In dataframe, it is `UnsafeRow` actually), it can avoid data copy. cc @liancheng @cloud-fan

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I will optimize `serialize`/`deserialize` by `ByteBuffer` later. Though it is not the bottleneck currently.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Support `weightCol` parameter in a convenient way.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Is there some better way to get the object of `VectorUDT` ?  cc @cloud-fan

File: mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala]: Performance test, on vector size from 1 to 10000. (This part code should be removed before the PR merged).

**Review by WeichenXu123 — COMMENTED** [review]
Made a pass by myself.

**Comment by thunterdb** [comment]
@WeichenXu123 thanks! Can you post some performance numbers as well?

**Comment by WeichenXu123** [comment]
performance data attached. cc @thunterdb @jkbradley

**Comment by thunterdb** [comment]
Thank you for the performance numbers @WeichenXu123 , I have a couple of comments:
 - you say that SQL uses adaptive compaction. How bad is that? I assume it adds some overhead.
 - did you just run each experiment once? I would be interested in error bars on these numbers, as it can take up to 30 seconds for the JVM to warm up and optimize the byte code. You should report the geometric mean or the median time of running these experiments to make sure that you are skewed by outliers. Some others will probably have some good advice as well.
 - from the performance numbers, there are 2 different regimes: small vectors, and big vectors (for which even the DataFrame -> RDD conversion is faster than working straight with DataFrames). I would be curious to know the bottlenecks for each case.

If we trust these numbers, the overall conclusion is that the SQL interface adds a 2x-3x performance overhead over RDDs for the time being. @cloud-fan @liancheng are there still some low hanging fruits that could be merged into SQL? 

This state of affair is of course far from great, but I am in favor of merging this piece and improve it iteratively with the help of the SQL team, as this code is easy to benchmark and representative of the rest of MLlib, once we start to rely more on dataframe and catalysts, and less on RDDs.

@yanboliang @viirya @kiszk what are your thoughts?

**Comment by thunterdb** [comment]
cc @hvanhovell as well.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: this is not going to be 2.2 anymore

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: please mention the type of the collection: `TraversableIndexedSeq[_]`

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: why do you need to do that?

Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I suspect this is going to cause some boxing on the values, too.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: private

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I am actually surprised you do not get a performance drop instead of using a vector or an array. Shouldn't it be private, too?

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Please add this comment to the source code itself.

File: mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala
Comment (thunterdb) [inline mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala]: Do not remove it, we will need it later. I think there is a way to tag the test as a performance harness, but I could not remember how. @liancheng , do you have some suggestions?

**Review by thunterdb — COMMENTED** [review]
@WeichenXu123 thanks a lot, I only have a few comments. Someone else should take a look, as I am the original author of this code.

**Comment by WeichenXu123** [comment]
@thunterdb 
1) The dataframe deserialize from binary data will add overhead, (maybe there is compaction or not, it depends on the datatype, cc @liancheng ) about 1x performance in my test.
2) My test try 20 times, the first 5 times used as warm up, the numbers in the table is the mean of the 15 times results. You can trust the result I think.
3) When the vector size is small (1~10), the bottleneck is at many small java object creation I think. So the performance result is different than the larger vector.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Yes... this method is only used in testsuite (let writing testing more convenient).

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I would like to ask whether we need make the whole `SummarizerBuffer` to be private?

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Here I make the `TraversableIndexedSeq vector` directly backend on the `UnsafeRow` from dataframe. (It acts like the `ByteBuffer` in java) avoid to copy the array data to generate a vector....

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: @WeichenXu123 Could you add comment that we operate on raw data (```UnsafeRow```) directly to avoid copy values?

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I'd like to make the whole ```SummarizerBuffer``` private, as users will not use it.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: If it's only used in testsuite, make comment more clear.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: +1 @thunterdb

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: It seems function ```serialize/deserialize``` only called once for each partition, so I agree it's not bottleneck.

File: mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala]: We can keep the performance test results as ```ignore("performance test")``` and update it when we get improvement. You can refer [what Spark SQL did](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala).

**Comment by yanboliang** [comment]
@WeichenXu123 @thunterdb 
Thanks for this great work, we are always happy to see improvement which can help us to migrate MLlib workload to Dataset based API.
Here are my two cents:
1, Form the performance test result, there are performance degradation from 3 -> 2 -> 1.
The difference between case 3 and case 2 is the deserialization cost of each instance, I suspect we can't skip this step, so it maybe more hard to optimize it. However, the difference between case 2 and case 1 is the aggregate implementation. The aggregate on Dataset will use user-defined type as aggregate buffer, so I think we should make some effort to improve ```TypedImperativeAggregate```. I guess it maybe not so hard compared with the previous bottleneck, but this is just my intuition, let’s check with SQL guys. cc @liancheng  @cloud-fan @hvanhovell 
2, @WeichenXu123 In the current experiment, I saw you only use two partitions. Is there any difference if we run against more partitions or more dataset?
3, @thunterdb I agree that we can get this in and make improvement continuously. When we get desirable result, we can start to migrate other MLlib workload on top of Dataset/catalysts. But we should add comment to let users know(if we can’t get desirable result before 
2.3.0) that the performance of Dataset-based multivariate statistic is not good enough currently. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Change to use `VectorUDT.deserialize` (Because last version code here won't improve performance markedly but increase code complexity.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Let's put a comment about performance to indicate that it is about 3x slower than using the RDD interface.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Put a comment about performance here.

**Review by thunterdb — COMMENTED** [review]
@WeichenXu123 thanks, this looks ready except for one comment about performance.

@yanboliang , can you merge this PR after this is addressed, if this looks good for you? Thank you in advance.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: I will trace this the performance gap here with related sql layer improvement in the future.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Comment by thunterdb** [comment]
@yanboliang do you feel comfortable to merge this PR? I think that all the questions have been addressed.

**Comment by yanboliang** [comment]
@thunterdb I'm on travel these days, will do a final pass and merge it on next Monday/Tuesday. Thanks.

**Comment by viirya** [comment]
Sorry can we make the performance data clear? Currently it doesn't say what the unit of the numbers is.

**Comment by WeichenXu123** [comment]
@viirya Sure! comment updated.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Remove this line?

**Comment by viirya** [comment]
@WeichenXu123 Thanks! Looks good.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: This `deserialize` on each `Vector` should be a bottleneck.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: From the usage of this `Vector`, I think we can directly work on the serialized Vector data (size, indices, values). It should reduce much of time on deserialization of `Vector`.

Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: @viirya I have tried your suggestion in the previous version code, but it do not bring performance advantage.
You can check my previous version code (in this commit "optimize summarizer buffer") and run tests.

File: mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala]: Remove them if it's not used.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Oh, I saw it. Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala]: Could you paste your performance test result here? Just like [here](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala).

**Comment by yanboliang** [comment]
@WeichenXu123 I left some minor comments, otherwise, LGTM. Thanks.

**Comment by WeichenXu123** [comment]
@yanboliang I will update ASAP, thanks!

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: If directly work on serialized data (UnsafeArrayData), it only avoid the array copy(which save little time), but brings extra cost when calling `UnsafeArrayData.getDouble`)

**Review by yanboliang — APPROVED** [review]
LGTM, I'll merge it tomorrow if there is no further discussion. Thanks for all.

File: mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: May I ask how the performance test runs? Especially for the RDD part.

Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: @viirya 
modify `ignore("performance test")` to `test("performance test")`
then run test: SummarizerSuite

Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala]: Thanks. I didn't review the test thoroughly.

**Comment by thunterdb** [comment]
Thank you @yanboliang.

**Comment by yanboliang** [comment]
Merged into master, thanks for all.

**Pull request closed without merging**
